### PR TITLE
System Status Tools API

### PR DESCRIPTION
**Pull request by emdashcodes**
This PR adds the following:
- System Stats Tools Endpoints: Get a list of tools, get info on a specific tool, and run/execute a tool.
- System Status Modes Endpoints: Allows you to toggle "modes" from the system status page (template debug, shipping debug, and uninstall data).
- Updates the system status page to use the tools controller so code is shared.
- Tests for all of the above.

### CODE REVIEW

File: includes/admin/class-wc-admin-status.php [hunk 1]
```diff
@@ -40,96 +40,13 @@ public static function status_tools() {
...
-						$callback = $tools[ $action ]['callback'];
-						$return = call_user_func( $callback );
-						if ( $return === false ) {
-							$callback_string = is_array( $callback ) ? get_class( $callback[0] ) . '::' . $callback[1] : $callback;
-							echo '<div class="error inline"><p>' . sprintf( __( 'There was an error calling %s', 'woocommerce' ), $callback_string ) . '</p></div>';
-						}
-					}
-				break;
+			if ( $response['success'] ) {
+				echo '<div class="updated inline"><p>' . $response['message'] . '</p></div>';
```
Comment (mikejolley) [inline includes/admin/class-wc-admin-status.php]: Missing escape

File: includes/admin/class-wc-admin-status.php [hunk 2]
```diff
@@ -40,96 +40,13 @@ public static function status_tools() {
...
+			} else {
+				echo '<div class="error inline"><p>' . $response['message'] . '</p></div>';
```
Comment (mikejolley) [inline includes/admin/class-wc-admin-status.php]: Missing escape

File: includes/admin/class-wc-admin-status.php [hunk 3]
```diff
@@ -40,96 +40,13 @@ public static function status_tools() {
 		$tools = self::get_tools();
 
 		if ( ! empty( $_GET['action'] ) && ! empty( $_REQUEST['_wpnonce'] ) && wp_verify_nonce( $_REQUEST['_wpnonce'], 'debug_action' ) ) {
+			$tools_controller = new WC_REST_System_Status_Tools_Controller;
+			$response = $tools_controller->execute_tool( $_GET['action'] );
```
Comment (mikejolley) [inline includes/admin/class-wc-admin-status.php]: Sanitize this

File: includes/api/class-wc-rest-system-status-controller.php [hunk 4]
```diff
@@ -86,6 +126,104 @@ public function get_items( $request ) {
 		return rest_ensure_response( $response );
 	}
 
+	/**
+	 * A list of modes that can be toggled via WC's system status screens
+	 *
+	 * @return array
+	 */
+	public function _get_modes() {
```
Comment (mikejolley) [inline includes/api/class-wc-rest-system-status-controller.php]: Totally forgot about modes. I wonder if we need these or if we can intelligently offer them elsewhere or via constants. Not sure if they should be enabled externally. Maybe we can discuss in team chat today?

File: includes/api/class-wc-rest-system-status-tools-controller.php [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+	 * Actually executes a a tool.
+	 *
+	 * @param  string $tool
+	 * @return array
+	 */
+	public function execute_tool( $tool ) {
+		global $wpdb;
+		$ran = true;
+		switch ( $tool ) {
+			case 'clear_transients' :
```
Comment (mikejolley) [inline includes/api/class-wc-rest-system-status-tools-controller.php]: Do we want to keep this naming or switch to hyphens?

**Comment by mikejolley** [comment]
Looks good. Added a few comments/questions above.

File: includes/api/class-wc-rest-system-status-tools-controller.php (hunk 5 above)
Comment (justinshreve) [inline includes/api/class-wc-rest-system-status-tools-controller.php]: We should keep underscores. All our API paths are currently using `_` and these slugs will be used in the path name for actually executing the tool.

**Comment by justinshreve** [comment]
Fixed & removed modes code. Will merge :).

**Pull request merged**